File: leetcodePractice/6-27-2020/tspWithPrunning.java
/*
tsp-剪枝 prunning
 */

 /*
tsp - java version
 */
public class Result {
    int minCost;
    public Result() {
        minCost = Integer.MAX_VALUE;
    }
}

public class Solution {

    public int misCost(int n, int[][] roads) {
        int[][] graph = constructGraph(roads, n);
        Set<Integer> visited = new HashSet<>();
        List<Integer> path = new ArrayList<>();
        Result result = new Result();
        visited.add(1);
        path.add(1);
        dfs(1, n, visited, path, 0, graph, result);
        

        return result.minCost;
    } 

    private void dfs(int city, 
                     int n, 
                     Set<Integer> visited, 
                     List<Integer> path,
                     int cost, 
                     int[][] graph,
                     Result result) {
        if (visited.size() == n) {
            result.minCost = Math.min(result.minCost, cost);
            return ;
        }

        for (int i = 1; i < graph[city].length; i++) {
            if (visited.contains(i)) {
                continue;
            }
            if (hasBetterPath(graph, path, city)) {
                continue;
            }

            visited.add(i); //代表可以走到i 城市
            path.add(i);
            dfs(i, n, visited, cost + graph[city][i], graph, result);
            visited.remove(i);
            path.remove(path.size() - 1);
        }
    }

    private boolean hasBetterPath(int[][] graph, List<Integer> path, int city) {
        for (int i = 0; i < path.length; i++) {
            int A = path.get(i - 1);
            int B = path.get(i);
            int last = path.get(path.length - 1);

            if (graph[A][B] + graph[last][city] > graph[A][last] + graph[B][city]) {
                return true;
            }
        }

        return false;
    }

    private int[][] constructGraph(int[][] roads, int n) {
        int[][] graph = new int[n + 1][n + 1];
        for (int i = 0; i < n + 1; i++) {
            for (int j = 0; j < n + 1; j++) {
                graph[i][j] = Integer.MAX_VALUE >> 4;
            }
        }
        for (int i = 0; i < roads.length; i++) {
            int a = roads[i][0];
            int b = roads[i][1];
            int c = roads[i][2];
            graph[a][b] = Math.min(graph[a][b], c);
            graph[b][a] = Math.min(graph[b][a], c);
        }
        return graph;
    }
}

File: leetcodePractice/6-27-2020/tsp_with_prunning.py
'''
最优化剪枝-python版本
tips：DFS尿性就是参数贼多，开始写没弄全很正常，发现了的话后面加上去就好
'''

class Result:
    def __init__(self):
        self.min_cost = float('inf')

# 以下都在 class Solution中
def minCost(self, n, roads):
    graph = self.construct_graph(roads, n)
    result = Result()
    self.dfs(1, n, [1], set([1]), 0, graph, result)
    return result.min_cost

def dfs(self, city, n, path, visited, cost, graph, result):
    if len(visited) == n:
        result.min_cost = min(result.min_cost, cost)
        return 
    
    for next_city in graph[city]:
        if next_city in visited:
            continue
        if self.has_better_path(graph, path, next_city):
            continue
        visited.add(next_city)
        path.append(next_city)
        self.dfs(
            next_city,
            n,
            path,
            visited,
            cost + graph[city][next_city],
            graph,
            result
        )
        path.pop()
        visited.remove(next_city)

def construct_graph(self, roads, n):
    graph = {
        i: {j: float('inf') for j in range(1, n + 1)}
        for i in range(1, n + 1)
    }
    for a, b, c in roads:
        graph[a][b] = min(graph[a][b], c)
        graph[b][a] = min(graph[b][a], c)
    return graph

# 如果感觉这个函数不好懂，就看下ipad里面的笔记，或者自己画个图
def has_better_path(self, graph, path, city):
    for i in range(1, len(path)):
        if graph[path[i - 1]][path[i]] + graph[path[-1]][city] >\
            graph[path[i - 1]][path[-1]] + graph[path[i]][city]:
            return True
        
    return False
